File: bindings/java-ffm/src/main/java/org/maplibre/nativeffi/runtime/RuntimeEventType.java
package org.maplibre.nativeffi.runtime;

/** Runtime event type copied from the native event queue. */
public final class RuntimeEventType {
  public static final RuntimeEventType MAP_CAMERA_WILL_CHANGE =
      new RuntimeEventType(1, "MAP_CAMERA_WILL_CHANGE");
  public static final RuntimeEventType MAP_CAMERA_IS_CHANGING =
      new RuntimeEventType(2, "MAP_CAMERA_IS_CHANGING");
  public static final RuntimeEventType MAP_CAMERA_DID_CHANGE =
      new RuntimeEventType(3, "MAP_CAMERA_DID_CHANGE");
  public static final RuntimeEventType MAP_STYLE_LOADED =
      new RuntimeEventType(4, "MAP_STYLE_LOADED");
  public static final RuntimeEventType MAP_LOADING_STARTED =
      new RuntimeEventType(5, "MAP_LOADING_STARTED");
  public static final RuntimeEventType MAP_LOADING_FINISHED =
      new RuntimeEventType(6, "MAP_LOADING_FINISHED");
  public static final RuntimeEventType MAP_LOADING_FAILED =
      new RuntimeEventType(7, "MAP_LOADING_FAILED");
  public static final RuntimeEventType MAP_IDLE = new RuntimeEventType(8, "MAP_IDLE");
  public static final RuntimeEventType MAP_RENDER_UPDATE_AVAILABLE =
      new RuntimeEventType(9, "MAP_RENDER_UPDATE_AVAILABLE");
  public static final RuntimeEventType MAP_RENDER_ERROR =
      new RuntimeEventType(10, "MAP_RENDER_ERROR");
  public static final RuntimeEventType MAP_STILL_IMAGE_FINISHED =
      new RuntimeEventType(11, "MAP_STILL_IMAGE_FINISHED");
  public static final RuntimeEventType MAP_STILL_IMAGE_FAILED =
      new RuntimeEventType(12, "MAP_STILL_IMAGE_FAILED");
  public static final RuntimeEventType MAP_RENDER_FRAME_STARTED =
      new RuntimeEventType(13, "MAP_RENDER_FRAME_STARTED");
  public static final RuntimeEventType MAP_RENDER_FRAME_FINISHED =
      new RuntimeEventType(14, "MAP_RENDER_FRAME_FINISHED");
  public static final RuntimeEventType MAP_RENDER_MAP_STARTED =
      new RuntimeEventType(15, "MAP_RENDER_MAP_STARTED");
  public static final RuntimeEventType MAP_RENDER_MAP_FINISHED =
      new RuntimeEventType(16, "MAP_RENDER_MAP_FINISHED");
  public static final RuntimeEventType MAP_STYLE_IMAGE_MISSING =
      new RuntimeEventType(17, "MAP_STYLE_IMAGE_MISSING");
  public static final RuntimeEventType MAP_TILE_ACTION =
      new RuntimeEventType(18, "MAP_TILE_ACTION");
  public static final RuntimeEventType OFFLINE_REGION_STATUS_CHANGED =
      new RuntimeEventType(19, "OFFLINE_REGION_STATUS_CHANGED");
  public static final RuntimeEventType OFFLINE_REGION_RESPONSE_ERROR =
      new RuntimeEventType(20, "OFFLINE_REGION_RESPONSE_ERROR");
  public static final RuntimeEventType OFFLINE_REGION_TILE_COUNT_LIMIT_EXCEEDED =
      new RuntimeEventType(21, "OFFLINE_REGION_TILE_COUNT_LIMIT_EXCEEDED");
  public static final RuntimeEventType OFFLINE_OPERATION_COMPLETED =
      new RuntimeEventType(22, "OFFLINE_OPERATION_COMPLETED");

  private final int nativeValue;
  private final String name;

  private RuntimeEventType(int nativeValue, String name) {
    this.nativeValue = nativeValue;
    this.name = name;
  }

  public int rawValue() {
    return nativeValue;
  }

  public static RuntimeEventType fromNative(int nativeValue) {
    return switch (nativeValue) {
      case 1 -> MAP_CAMERA_WILL_CHANGE;
      case 2 -> MAP_CAMERA_IS_CHANGING;
      case 3 -> MAP_CAMERA_DID_CHANGE;
      case 4 -> MAP_STYLE_LOADED;
      case 5 -> MAP_LOADING_STARTED;
      case 6 -> MAP_LOADING_FINISHED;
      case 7 -> MAP_LOADING_FAILED;
      case 8 -> MAP_IDLE;
      case 9 -> MAP_RENDER_UPDATE_AVAILABLE;
      case 10 -> MAP_RENDER_ERROR;
      case 11 -> MAP_STILL_IMAGE_FINISHED;
      case 12 -> MAP_STILL_IMAGE_FAILED;
      case 13 -> MAP_RENDER_FRAME_STARTED;
      case 14 -> MAP_RENDER_FRAME_FINISHED;
      case 15 -> MAP_RENDER_MAP_STARTED;
      case 16 -> MAP_RENDER_MAP_FINISHED;
      case 17 -> MAP_STYLE_IMAGE_MISSING;
      case 18 -> MAP_TILE_ACTION;
      case 19 -> OFFLINE_REGION_STATUS_CHANGED;
      case 20 -> OFFLINE_REGION_RESPONSE_ERROR;
      case 21 -> OFFLINE_REGION_TILE_COUNT_LIMIT_EXCEEDED;
      case 22 -> OFFLINE_OPERATION_COMPLETED;
      default -> new RuntimeEventType(nativeValue, null);
    };
  }

  @Override
  public boolean equals(Object other) {
    return other instanceof RuntimeEventType value && nativeValue == value.nativeValue;
  }

  @Override
  public int hashCode() {
    return Integer.hashCode(nativeValue);
  }

  @Override
  public String toString() {
    return name != null ? name : "RuntimeEventType(" + nativeValue + ")";
  }
}

File: bindings/kotlin/src/commonMain/kotlin/org/maplibre/nativeffi/runtime/RuntimeEventType.kt
package org.maplibre.nativeffi.runtime

import kotlin.jvm.JvmInline

/**
 * Runtime event type copied from the native event queue. It selects the meaning of
 * [RuntimeEvent.code] and the [RuntimeEventPayload] variant behind [RuntimeEvent.payload].
 *
 * This is an open domain: a value may have no named constant here, so a `when` over this type needs
 * an `else` branch.
 */
@JvmInline
public value class RuntimeEventType(public val nativeValue: Int) {
  public companion object {
    /** [RuntimeEvent.code] carries a [CameraChangeMode]. */
    public val MAP_CAMERA_WILL_CHANGE: RuntimeEventType = RuntimeEventType(1)

    public val MAP_CAMERA_IS_CHANGING: RuntimeEventType = RuntimeEventType(2)

    /** [RuntimeEvent.code] carries a [CameraChangeMode]. */
    public val MAP_CAMERA_DID_CHANGE: RuntimeEventType = RuntimeEventType(3)

    public val MAP_STYLE_LOADED: RuntimeEventType = RuntimeEventType(4)
    public val MAP_LOADING_STARTED: RuntimeEventType = RuntimeEventType(5)
    public val MAP_LOADING_FINISHED: RuntimeEventType = RuntimeEventType(6)

    /**
     * [RuntimeEvent.code] carries the ordinal of MapLibre Native's internal map load error kind,
     * which the C API leaves unnamed. [RuntimeEvent.message] carries the failure text.
     */
    public val MAP_LOADING_FAILED: RuntimeEventType = RuntimeEventType(7)

    public val MAP_IDLE: RuntimeEventType = RuntimeEventType(8)
    public val MAP_RENDER_UPDATE_AVAILABLE: RuntimeEventType = RuntimeEventType(9)
    public val MAP_RENDER_ERROR: RuntimeEventType = RuntimeEventType(10)
    public val MAP_STILL_IMAGE_FINISHED: RuntimeEventType = RuntimeEventType(11)
    public val MAP_STILL_IMAGE_FAILED: RuntimeEventType = RuntimeEventType(12)
    public val MAP_RENDER_FRAME_STARTED: RuntimeEventType = RuntimeEventType(13)
    public val MAP_RENDER_FRAME_FINISHED: RuntimeEventType = RuntimeEventType(14)
    public val MAP_RENDER_MAP_STARTED: RuntimeEventType = RuntimeEventType(15)
    public val MAP_RENDER_MAP_FINISHED: RuntimeEventType = RuntimeEventType(16)

    /** [RuntimeEvent.message] carries the missing image id. */
    public val MAP_STYLE_IMAGE_MISSING: RuntimeEventType = RuntimeEventType(17)

    /** [RuntimeEvent.message] carries the source id. */
    public val MAP_TILE_ACTION: RuntimeEventType = RuntimeEventType(18)

    public val OFFLINE_REGION_STATUS_CHANGED: RuntimeEventType = RuntimeEventType(19)
    public val OFFLINE_REGION_RESPONSE_ERROR: RuntimeEventType = RuntimeEventType(20)
    public val OFFLINE_REGION_TILE_COUNT_LIMIT_EXCEEDED: RuntimeEventType = RuntimeEventType(21)

    /**
     * [RuntimeEvent.code] carries the operation result as a native status value, the same value
     * [RuntimeEventPayload.OfflineOperationCompleted.resultStatus] reports.
     */
    public val OFFLINE_OPERATION_COMPLETED: RuntimeEventType = RuntimeEventType(22)

    /** [RuntimeEvent.payload] carries [RuntimeEventPayload.CameraTransitionFinished]. */
    public val MAP_CAMERA_TRANSITION_FINISHED: RuntimeEventType = RuntimeEventType(23)

    internal fun fromNative(nativeValue: UInt): RuntimeEventType = fromNative(nativeValue.toInt())

    internal fun fromNative(nativeValue: Int): RuntimeEventType = RuntimeEventType(nativeValue)
  }
}
